internal/sync/lifecycle.go

```diff
@@ -117,39 +101,45 @@ func Once() LifecycleOnce {
 // If Start is called multiple times it will return the error
 // from the first time it was called.
 func (l *lifecycleOnce) Start(f func() error) error {
-	if l.starting.Swap(true) {
-		<-l.startCh
+	if l.state.CAS(int32(Idle), int32(Starting)) {
+		if f != nil {
+			l.err = f()
+		}
+
+		// skip forward to error state
+		if l.err != nil {
+			l.state.Store(int32(Errored))
+			close(l.stopCh)
+		} else {
+			l.state.Store(int32(Running))
+		}
+		close(l.startCh)
 		return l.err
 	}
-	if f != nil {
-		l.err = f()
-	}
-	// skip forward to error state
-	if l.err != nil {
-		l.errored.Store(true)
-		l.stopped.Store(true)
-		l.stopping.Store(true)
-		close(l.stopCh)
-	}
-	l.started.Store(true)
-	close(l.startCh)
 
+	<-l.startCh
 	return l.err
 }
 
 func (l *lifecycleOnce) WhenRunning(ctx context.Context) error {
-	if !l.stopping.Load() && l.started.Load() {
+	state := LifecycleState(l.state.Load())
+	if state == Running {
 		return nil
 	}
+	if state > Running {
+		return context.DeadlineExceeded
+	}
 
 	if _, ok := ctx.Deadline(); !ok {
 		return errDeadlineRequired
 	}
 
 	select {
 	case <-l.startCh:
-		return nil
-	case <-l.stopCh:
+		state := LifecycleState(l.state.Load())
+		if state == Running {
+			return nil
+		}
```

Comment:
We may need to revisit all our places we propagate deadline exceeded and wrap them with more context about what part of the YARPC flow they timed out in. We may need a deadline exceeded error "behavior" interface à la Dave Cheney, and wrapper types that provide more descriptive feedback about who had the hot potato.